HOSTEDCP-965: Add impersonate feature to the CLI and document HC dump procedure (#2653)
openshift/hypershift: docs/content/how-to/troubleshooting.md

Reviewer: Do we or can we add more examples on what can be passed in for the `--as` flag?

Author: Yep, makes sense, I will add another one using a cluster user. Anything more that could you have in mind?

Reviewer: No, not really. Just anything that would be useful for us to point users to look at.

# Troubleshooting

## How to dump the HostedCluster resources from a management cluster

In order to dump the relevant HostedCluster objects we will need some prerequisites:

- `cluster-admin` access to the management cluster
- The HostedCluster `name` and the `namespace` where the CR is deployed
- The Hypershift CLI

> - Th Openshift CLI

Reviewer 2: s/Th/The/

# Troubleshooting

## How to dump the HostedCluster resources from a management cluster

In order to dump the relevant HostedCluster objects we will need some prerequisites:

- `cluster-admin` access to the management cluster
- The HostedCluster `name` and the `namespace` where the CR is deployed
- The Hypershift CLI

Once we have these elements and the shell with the Kubeconfig loaded pointing to the management cluster, we will execute these commands:

```bash
CLUSTERNAME="samplecluster"
CLUSTERNS="clusters"

mkdir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
hypershift dump cluster \
    --name ${CLUSTERNAME} \
    --namespace ${CLUSTERNS} \
    --artifact-dir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
```

After some time, the output will show something like this:

```bash
2023-06-06T12:18:20+02:00	INFO	Archiving dump	{"command": "tar", "args": ["-cvzf", "hypershift-dump.tar.gz", "cluster-scoped-resources", "event-filter.html", "namespaces", "network_logs", "timestamp"]}
2023-06-06T12:18:21+02:00	INFO	Successfully archived dump	{"duration": "1.519376292s"}
```

This dump contains so much and quite useful information that will help on the troubleshooting procedure.

### Impersonation

> The dump command could be use with the flag `--as` which works in the same way as the `oc` client. If you executes the command with the flag, the CLI will impersonate all the queries against the management cluster, to use that username or service account.

Reviewer: ```suggestion
The dump command can be used with the flag `--as`, which works in the same way as the `oc` client. If you execute the command with the flag, the CLI will impersonate all the queries against the management cluster, using that username or service account.
```

> The service account should has enough permissions to query all the objects from some namespaces, so cluster-admin is recommended to make sure you have enough permissions. That service account should be located (or have permissions to query) al least in the HostedControlPlane namespace.

Reviewer: ```suggestion
The service account should have enough permissions to query all the objects from the namespaces, so cluster-admin is recommended to make sure you have enough permissions. The service account should be located (or have permissions to query) at least the HostedControlPlane namespace.
```

# Troubleshooting

## How to dump the HostedCluster resources from a management cluster

In order to dump the relevant HostedCluster objects we will need some prerequisites:

- `cluster-admin` access to the management cluster
- The HostedCluster `name` and the `namespace` where the CR is deployed
- The Hypershift CLI

Once we have these elements and the shell with the Kubeconfig loaded pointing to the management cluster, we will execute these commands:

```bash
CLUSTERNAME="samplecluster"
CLUSTERNS="clusters"

mkdir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
hypershift dump cluster \
    --name ${CLUSTERNAME} \

>     --namespace ${CLUSTERNS} \

Reviewer 2: We should include --dump-guest-cluster given that it's desirable most of the time.

    --artifact-dir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
```

After some time, the output will show something like this:

```bash
2023-06-06T12:18:20+02:00	INFO	Archiving dump	{"command": "tar", "args": ["-cvzf", "hypershift-dump.tar.gz", "cluster-scoped-resources", "event-filter.html", "namespaces", "network_logs", "timestamp"]}
2023-06-06T12:18:21+02:00	INFO	Successfully archived dump	{"duration": "1.519376292s"}
```

> This dump contains so much and quite useful information that will help on the troubleshooting procedure.

Reviewer 2: s/This dump contains so much and quite useful information that will help on the troubleshooting procedure./This dump contains artifacts that aid in troubleshooting issues with hosted control plane clusters./

### Impersonation

The dump command can be used with the flag `--as`, which works in the same way as the `oc` client. If you execute the command with the flag, the CLI will impersonate all the queries against the management cluster, using that username or service account.

The service account should have enough permissions to query all the objects from the namespaces, so cluster-admin is recommended to make sure you have enough permissions. The service account should be located (or have permissions to query) at least the HostedControlPlane namespace.

!!! note

    If your user/sa doesn't have enough permissions, the command will be executed and will dump only the objects you have permissions to get and during that process some `forbidden` errors will be raised.

- Impersonation Sample using a service account:

```bash
CLUSTERNAME="samplecluster"
CLUSTERNS="clusters"
SA="samplesa"

mkdir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
hypershift dump cluster \
    --name ${CLUSTERNAME} \

>     --namespace ${CLUSTERNS} \

Reviewer 2: include `--dump-guest-cluster`

>     --as "system:serviceaccount:${CLUSTERNS}-${CLUSTERNAME}:${SA}" \

Reviewer 2: Instead of ${CLUSTERNS}-${CLUSTERNAME} I would have a separate variable like `${SA_NAMESPACE}`
The service account doesn't need to live in the control plane namespace.

    --artifact-dir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
```

- Impersonation Sample using a user:

```bash
CLUSTERNAME="samplecluster"
CLUSTERNS="clusters"
CLUSTERUSER="cloud-admin"

mkdir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
hypershift dump cluster \
    --name ${CLUSTERNAME} \

>     --namespace ${CLUSTERNS} \

Reviewer 2: include `--dump-guest-cluster`

    --as "${CLUSTERUSER}" \
    --artifact-dir clusterDump-${CLUSTERNS}-${CLUSTERNAME}
```

### Dump Content

The dump contains:

> - **Cluster-wide resources**: Like nodes definitions of the management cluster.

Reviewer 3: can we categorise this section in management cluster and guest cluster?

Reviewer 3: e.g mangement cluster
- Hypershift ns
- hypershift.io API resources
- Control plane namespace resources
Guest cluster
- ...

No need to be super granular
